### ISSUE TITLE
what's the difference between `say` and `respond`

### ISSUE THREAD

**Issue by imdgr886** [slackapi/bolt-python#973]
I know I can response the action hook with either `say` or `respond`, but what's the difference

https://slack.dev/bolt-python/concepts#action-listening

**Comment by seratch** [comment]
Hi @imdgr886, thanks for asking the quesiton. 

TL;DR -- If you're happy with `say()`, you don't need to worry about `respond()`!

* The `say` utility in Bolt: performs [`chat.postMessage` API](https://api.slack.com/methods/chat.postMessage) calls using the channel ID in a payload
* The `respond` utility in Bolt: performs an HTTP request with `response_url`, which can be included in a payload (a kind of incoming webhook associated with an interaction context such as user and channel); Perhaps, you may not be familiar with `response_url`. Read https://api.slack.com/interactivity/handling#message_responses to understand what it is.

Actually, both `respond` and `say` send a message in the channel where the user interaction happens, but there are a few differences:

* You can override the channel ID for `say` (chat.postMessage) while you cannot for `respond` (response_url)
* You can send either in_channel or ephemeral message with `respond` while you can send only in_channel one with `say` (if you want to send an ephemeral one, you can use `chat.postEphemeral` instead)
* You can replace/update an ephemeral message with `respond` while you cannot do the same with `chat.postEphemeral` API
* You can use the same response_url (aka `respond`) only three times while you call `say` (chat.postMessage) many times

If these sound confusing to you, you can simply use `say()` to post a message! I hope this clarifies.

**Comment by seratch** [comment]
Thanks for 👍 reaction! Let me close this issue now but whenever you have follow-up questions, please feel free to write in.
